test/extensions/transport_sockets/tls/handshaker_test.cc

```diff
@@ -0,0 +1,256 @@
+  EXPECT_CALL(handshaker_callbacks, onSuccessCb).Times(1);
+
+  HandshakerImpl handshaker(std::move(server_ssl_));
+
+  handshaker.setCallbacks(transport_socket_callbacks, handshaker_callbacks);
+
+  auto socket_state = Ssl::SocketState::PreHandshake;
+  auto post_io_action = Network::PostIoAction::KeepOpen; // default enum
+
+  // Run the handshakes from the client and server until HandshakerImpl decides
+  // we're done and returns PostIoAction::Close.
+  while (post_io_action != Network::PostIoAction::Close) {
+    SSL_do_handshake(client_ssl_.get());
+    post_io_action = handshaker.doHandshake(socket_state);
+  }
+
+  EXPECT_EQ(post_io_action, Network::PostIoAction::Close);
+  // HandshakerImpl should have set |socket_state| accordingly.
+  EXPECT_EQ(socket_state, Ssl::SocketState::HandshakeComplete);
+}
+
+// We induce some kind of BIO mismatch and force the SSL_do_handshake to
+// return an error code without error handling, i.e. not SSL_ERROR_WANT_READ
+// or _WRITE or _PRIVATE_KEY_OPERATION.
+TEST_F(HandshakerTest, ErrorCbOnAbnormalOperation) {
+  // We make a new BIO, set it as the `rbio`/`wbio` for the client SSL object,
+  // and break the BIO pair connecting the two SSL objects. Now handshaking will
+  // fail, likely with SSL_ERROR_SSL.
+  BIO* bio = BIO_new(BIO_s_socket());
+  SSL_set_bio(client_ssl_.get(), bio, bio);
+
+  HandshakerImpl handshaker(std::move(server_ssl_));
+
+  StrictMock<Network::MockTransportSocketCallbacks> transport_socket_callbacks;
+
+  StrictMock<MockHandshakerCallbacks> handshaker_callbacks;
+  EXPECT_CALL(handshaker_callbacks, onFailureCb).Times(1);
+
+  handshaker.setCallbacks(transport_socket_callbacks, handshaker_callbacks);
+
+  auto socket_state = Ssl::SocketState::PreHandshake;
+  auto post_io_action = Network::PostIoAction::KeepOpen; // default enum
+
+  while (post_io_action != Network::PostIoAction::Close) {
+    SSL_do_handshake(client_ssl_.get());
+    post_io_action = handshaker.doHandshake(socket_state);
+  }
+
+  // In the error case, HandshakerImpl also closes the connection.
+  EXPECT_EQ(post_io_action, Network::PostIoAction::Close);
+}
+
+// Example HandshakerImpl demonstrating special-case behavior which necessitates
+// extra SSL_ERROR case handling. Here, we induce an SSL_ERROR_WANT_X509_LOOKUP,
+// check for it in the handshaker, faux-trigger the lookup, and then proceed as
+// normal.
+class HandshakerImplForTest : public Ssl::Handshaker {
+public:
+  HandshakerImplForTest(bssl::UniquePtr<SSL> ssl, std::function<void()> requested_cert_cb)
+      : ssl_(std::move(ssl)), requested_cert_cb_(requested_cert_cb) {
+    SSL_set_cert_cb(
+        ssl_.get(), [](SSL*, void* arg) -> int { return *static_cast<bool*>(arg) ? 1 : -1; },
+        &cert_cb_ok_);
+  }
+
+  Network::PostIoAction doHandshake(Ssl::SocketState& state) override {
+    ASSERT(state != Ssl::SocketState::HandshakeComplete && state != Ssl::SocketState::ShutdownSent);
```

Comment:
This for test doHandshake implementation looks too real.  it seems like a copy of HandshakeImpl::doHandshake.

Reply:
It basically is. I specifically wanted to demonstrate special-case error handling under `switch(SSL_get_error(...))`. Should I add more comments to call out that only the `SSL_ERROR_WANT_X509_LOOKUP` block is different?